### PR TITLE
[SPARK-30409][SPARK-29173][SQL][TESTS] Use `NoOp` datasource in SQL benchmarks

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to replace `.collect()`, `.count()` and `.foreach(_ => ())` in SQL benchmarks and use the `NoOp` datasource. I added an implicit class to `SqlBasedBenchmark` with the `.noop()` method. It can be used in benchmark like: `ds.noop()`. The last one is unfolded to `ds.write.format("noop").mode(Overwrite).save()`. 

### Why are the changes needed?
To avoid additional overhead that `collect()` (and other actions) has. For example, `.collect()` has to convert values according to external types and pull data to the driver. This can hide actual performance regressions or improvements of benchmarked operations.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Re-run all modified benchmarks using Amazon EC2.

| Item | Description |
| ---- | ----|
| Region | us-west-2 (Oregon) |
| Instance | r3.xlarge (spot instance) |
| AMI | ami-06f2f779464715dc5 (ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-20190722.1) |
| Java | OpenJDK8/11 |

- Run `TPCDSQueryBenchmark` using instructions from the PR #26049
```
# `spark-tpcds-datagen` needs this. (JDK8)
$ git clone https://github.com/apache/spark.git -b branch-2.4 --depth 1 spark-2.4
$ export SPARK_HOME=$PWD
$ ./build/mvn clean package -DskipTests

# Generate data. (JDK8)
$ git clone git@github.com:maropu/spark-tpcds-datagen.git
$ cd spark-tpcds-datagen/
$ build/mvn clean package
$ mkdir -p /data/tpcds
$ ./bin/dsdgen --output-location /data/tpcds/s1  // This need `Spark 2.4`
```
- Other benchmarks ran by the script:
```
#!/usr/bin/env python3

import os
from sparktestsupport.shellutils import run_cmd

benchmarks = [
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.AggregateBenchmark'],
    ['avro/test', 'org.apache.spark.sql.execution.benchmark.AvroReadBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.BloomFilterBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.DataSourceReadBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.DateTimeBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.ExtractBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.FilterPushdownBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.InExpressionBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.IntervalBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.JoinBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.MakeDateTimeBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.MiscBenchmark'],
    ['hive/test', 'org.apache.spark.sql.execution.benchmark.ObjectHashAggregateExecBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.OrcNestedSchemaPruningBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.OrcV2NestedSchemaPruningBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.ParquetNestedSchemaPruningBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.RangeBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.UDFBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.WideSchemaBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.benchmark.WideTableBenchmark'],
    ['hive/test', 'org.apache.spark.sql.hive.orc.OrcReadBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.datasources.csv.CSVBenchmark'],
    ['sql/test', 'org.apache.spark.sql.execution.datasources.json.JsonBenchmark']
]

print('Set SPARK_GENERATE_BENCHMARK_FILES=1')
os.environ['SPARK_GENERATE_BENCHMARK_FILES'] = '1'

for b in benchmarks:
    print("Run benchmark: %s" % b[1])
    run_cmd(['build/sbt', '%s:runMain %s' % (b[0], b[1])])
```

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
Looks reasonable to me if that's the best way to execute without doing anything else

**Comment by MaxGekk** [comment]
> ... if that's the best way to execute without doing anything else

NoOp datasource just materializes rows and doesn't do anything else.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Should we update the benchmark results too?

**Comment by MaxGekk** [comment]
> Should we update the benchmark results too?

I can run all SQL benchmarks locally on Mac OS or I can launch an EC2 instance with Linux (not sure how much will it cost to me). My concerns are only:
1. New benchmark results will significantly increase PR size
2. @dongjoon-hyun usually run all benchmarks on particular configuration. I don't know how to reproduce the same environment at my side.

File: dev/run-benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+from sparktestsupport.shellutils import run_cmd
+
+benchmarks = [
```
Comment (HyukjinKwon) [inline dev/run-benchmarks.py]: Hey, this is nice but why dont you make a separate PR for this file alone :).

Comment (MaxGekk) [reply dev/run-benchmarks.py]: I am running all those benchmarks on a separate EC2 instance. As soon as it is done (running whole night already). I will commit results and revert the script from the PR.

**Comment by dongjoon-hyun** [comment]
@MaxGekk . Please see the commit log for the following.
> @dongjoon-hyun usually run all benchmarks on particular configuration.

For example, the followings.
- cb501771fad350a3f165354629f66c4a5e61557c ([SPARK-25668][SQL][TESTS] Refactor TPCDSQueryBenchmark to use main method)
- 854a0f752e3125dc9bce6b91c80d6a7f534e9389 ([SPARK-29320][TESTS] Compare `sql/core` module in JDK8/11 (Part 1))

It's always `r3.xlarge`. And, previously, I used CentOS. And after your request on kernel update, I used Ubuntu. I'd like to recommend you to update the PR description about your environment.

**Comment by srowen** [comment]
OK, while the benchmark changes now reflect a different instance type, maybe it's a fine time to reset this baseline to Java 11, a more modern arch, and because of the no-op change here.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@HyukjinKwon @srowen @dongjoon-hyun I think the PR is ready for reviews. Please, take a look at it one more time.

**Comment by dongjoon-hyun** [comment]
Thank you, @MaxGekk . I'll do.

File: sql/core/benchmarks/AggregateBenchmark-results.txt [hunk 2]
```diff
@@ -2,142 +2,142 @@
...
 
-OpenJDK 64-Bit Server VM 1.8.0_222-b10 on Linux 3.10.0-862.3.2.el7.x86_64
-Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
+OpenJDK 64-Bit Server VM 1.8.0_232-8u232-b09-0ubuntu1~18.04.1-b09 on Linux 4.15.0-1051-aws
+Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz
 stddev:                                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-stddev wholestage off                             13393          13403          14          7.8         127.7       1.0X
-stddev wholestage on                               1291           1315          25         81.2          12.3      10.4X
+stddev wholestage off                              6232           6243          15         16.8          59.4       1.0X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/AggregateBenchmark-results.txt]: With the change, it seems that JDK8 result become faster 2x+ (while JDK11 does not)
Do you have any idea about this impressive improvement on JDK8?
cc @dbtsai

File: external/avro/benchmarks/AvroReadBenchmark-jdk11-results.txt [hunk 3]
```diff
@@ -2,121 +2,121 @@
 SQL Single Numeric Column Scan
 ================================================================================================
 
-OpenJDK 64-Bit Server VM 11.0.4+11-LTS on Linux 3.10.0-862.3.2.el7.x86_64
-Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
+Java HotSpot(TM) 64-Bit Server VM 11.0.5+10-LTS on Linux 4.15.0-1051-aws
```
Comment (dongjoon-hyun) [inline external/avro/benchmarks/AvroReadBenchmark-jdk11-results.txt]: Ur, is this not `OpenJDK`? For JDK8 result, this PR is using `OpenJDK 64-Bit Server VM 1.8.0_232-8u232-b09-0ubuntu1~18.04.1-b09`.

**Review by dongjoon-hyun — COMMENTED** [review]
@MaxGekk . This PR is very useful because your patch reduces the benchmark running time and updates the benchmark result.

However, it's a little difficult to review because this PR doesn't change one factor. In other words, this PR changes multiple factors.
1. The benchmark target: The changed benchmark code.
2. The benchmark machine: `r3.xlarge` -> `m5.4xlarge`. (Is there a reason to change the environment? Initially, `r3.xlarge` is used before according to Databricks benchmark environment and to use the instance store disk for disk oriented benchmarks)
3. Different JVM? (I guess you can install and use OpenJDK for both Java8/Java11 although it doesn't matter much)

In any way, I support this PR and will run a benchmark from my side tomorrow to give more credential to your PR. Then, we can merge this PR as a base-line from now.

To reviewers, let's hold on this PR for a few days more.

**Comment by dongjoon-hyun** [comment]
BTW, I have two questions.
1. You don't have any additional settings like SBT_OPT, right?
2. In the EC2 machine, how do you run the runner script described in the PR description? 
```
#!/usr/bin/env python3

import os
from sparktestsupport.shellutils import run_cmd
...
```

**Comment by dongjoon-hyun** [comment]
Also, cc @dbtsai since he is interested in this comparison.

File: external/avro/benchmarks/AvroReadBenchmark-jdk11-results.txt (hunk 3 above)
Comment (MaxGekk) [reply external/avro/benchmarks/AvroReadBenchmark-jdk11-results.txt]: I took JDK 8 from Ubuntu 18.04 but JDK 11 from Oracle website: https://www.oracle.com/technetwork/java/javase/downloads/jdk11-downloads-5066655.html

**Comment by MaxGekk** [comment]
> r3.xlarge -> m5.4xlarge. (Is there a reason to change the environment?

m5.4xlarge is available for me for "free" since it has been already paid. To run on r3.xlarge, I had to pay additional costs for a couple days.

> Different JVM? (I guess you can install and use OpenJDK for both Java8/Java11 although it doesn't matter much)

Frankly speaking, I didn't expect that Oracle's JDK 11 is significantly different from OpenJDK 11. Next time, I will pay more attention for that.

> You don't have any additional settings like SBT_OPT, right?

I have this:
```
export MAVEN_OPTS="-Xmx4g -XX:MaxPermSize=512M -XX:ReservedCodeCacheSize=512m"
export SBT_OPTS="-Xmx4g -XX:MaxPermSize=512M -XX:ReservedCodeCacheSize=512m"
```

**Comment by dongjoon-hyun** [comment]
Are you using a public AMI? The one in the PR description, `ami-1dab2163`, looks weird to me. I tried to search your AMI, but I cannot find via search. The ID seems to be too short.

> Re-run all modified benchmarks using Amazon EC2 m5.4xlarge on Ubuntu 18.04.3 LTS (ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-20191002 (ami-1dab2163)) 

And, if I search with the name, the similar one is `ami-06d51e91cea0dac8d` (with different ID.)
Screen Shot 2020-01-09 at 10 56 37 PM

**Comment by MaxGekk** [comment]
@dongjoon-hyun The AMI exists in the **eu-north-1** region - Europe (Stockholm):
Screen Shot 2020-01-10 at 10 10 27

**Comment by dongjoon-hyun** [comment]
Oh. That's the difference. I always used `Oregon` region. Thanks.

**Comment by dongjoon-hyun** [comment]
Hi, @MaxGekk . In order to focus on `NoOp` change only, I regenerate the result on the r3.xlarge with OpenJDK8/11 with the default memory. Could you review and merge this PR?
- https://github.com/MaxGekk/spark/pull/24

**Comment by dongjoon-hyun** [comment]
Here is TPCDS result, @MaxGekk .
- https://github.com/MaxGekk/spark/pull/25

**Comment by MaxGekk** [comment]
@dongjoon-hyun  oh, sorry I missed your second PR.

**Comment by dongjoon-hyun** [comment]
Thank you. No problem, @MaxGekk .

What I want to say is that this `NoOp` PR improves many cases according to the design, but some cases show a regression in the number of execution time. I believe it's not due to the change of this PR. It would be the change on the underlying master branch code.

**Comment by MaxGekk** [comment]
> It would be the change on the underlying master branch code.

I think it makes sense to open separate JIRAs per each case as we did before?

File: sql/core/benchmarks/JoinBenchmark-results.txt [hunk 4]
```diff
@@ -2,74 +2,74 @@
...
 
-OpenJDK 64-Bit Server VM 1.8.0_222-b10 on Linux 3.10.0-862.3.2.el7.x86_64
+OpenJDK 64-Bit Server VM 1.8.0_232-8u232-b09-0ubuntu1~18.04.1-b09 on Linux 4.15.0-1044-aws
 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
 Join w long:                              Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-Join w long wholestage off                         4685           4814         182          4.5         223.4       1.0X
-Join w long wholestage on                           440            524         102         47.7          21.0      10.7X
+Join w long wholestage off                         4531           4557          37          4.6         216.1       1.0X
+Join w long wholestage on                          1214           1310          95         17.3          57.9       3.7X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/JoinBenchmark-results.txt]: @MaxGekk . For example, this case, this slowed down 3x times. It's a huge regression.

Comment (dongjoon-hyun) [reply sql/core/benchmarks/JoinBenchmark-results.txt]: Could you file a JIRA issue for this `JoinBenchmark`?

Comment (MaxGekk) [reply sql/core/benchmarks/JoinBenchmark-results.txt]: Sure, I will do that tomorrow.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master. Thank you, @MaxGekk and all.

File: sql/core/benchmarks/IntervalBenchmark-results.txt [hunk 5]
```diff
@@ -1,29 +1,29 @@
...
+7 units w/o interval                               2334           2336           2          0.4        2334.1       0.3X
+8 units w/ interval                                2533           2546          16          0.4        2533.0       0.3X
+8 units w/o interval                               2519           2521           1          0.4        2519.4       0.3X
+9 units w/ interval                                2885           2889           5          0.3        2884.5       0.2X
+9 units w/o interval                               2804           2813          12          0.4        2803.9       0.2X
+10 units w/ interval                               3041           3060          16          0.3        3041.3       0.2X
+10 units w/o interval                              3031           3043          15          0.3        3031.2       0.2X
+11 units w/ interval                               3270           3280           9          0.3        3269.9       0.2X
+11 units w/o interval                              3273           3280           7          0.3        3272.6       0.2X
```
Comment (MaxGekk) [inline sql/core/benchmarks/IntervalBenchmark-results.txt]: Regression by 66%. Here is the JIRA for investigation: https://issues.apache.org/jira/browse/SPARK-30562

File: sql/core/benchmarks/JoinBenchmark-results.txt (hunk 4 above)
Comment (MaxGekk) [reply sql/core/benchmarks/JoinBenchmark-results.txt]: Here is the JIRA: https://issues.apache.org/jira/browse/SPARK-30563

File: sql/core/benchmarks/JsonBenchmark-results.txt [hunk 6]
```diff
@@ -3,110 +3,110 @@ Benchmark for performance of JSON parsing
...
 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
 Json files in the per-line mode:          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-Text read                                         10782          10801          18          4.6         215.6       1.0X
-Schema inferring                                  76817          77205         623          0.7        1536.3       0.1X
-Parsing without charset                           90638          91395         794          0.6        1812.8       0.1X
-Parsing with UTF-8                               120085         121975        1705          0.4        2401.7       0.1X
+Text read                                         10249          10314          77          4.9         205.0       1.0X
+Schema inferring                                  35403          35436          40          1.4         708.1       0.3X
+Parsing without charset                           32875          32879           4          1.5         657.5       0.3X
```
Comment (MaxGekk) [inline sql/core/benchmarks/JsonBenchmark-results.txt]: This became faster almost up to 3 times

File: sql/core/benchmarks/WideSchemaBenchmark-results.txt [hunk 7]
```diff
@@ -2,144 +2,144 @@
...
-parsing large select:                    Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------
-1 select expressions                             6 /   13          0.0     5997373.0       1.0X
-100 select expressions                           7 /   10          0.0     7204596.0       0.8X
-2500 select expressions                        103 /  107          0.0   102962705.0       0.1X
+parsing large select:                     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+1 select expressions                                  5             13           8          0.0     5370143.0       1.0X
+100 select expressions                               12             16           6          0.0    11995425.0       0.4X
+2500 select expressions                             211            214           4          0.0   210927791.0       0.0X
```
Comment (MaxGekk) [inline sql/core/benchmarks/WideSchemaBenchmark-results.txt]: Here there is slow down ~2 times. I opened JIRA for investigations: https://issues.apache.org/jira/browse/SPARK-30564

File: sql/core/benchmarks/WideSchemaBenchmark-results.txt [hunk 8]
```diff
@@ -2,144 +2,144 @@
...
+1 deep x 100000 rows (read in-mem)                   37             41           6          2.7         374.5       1.0X
+1 deep x 100000 rows (exec in-mem)                   47             50           6          2.1         466.9       0.8X
+1 deep x 100000 rows (read parquet)                  58             61           7          1.7         577.7       0.6X
+1 deep x 100000 rows (write parquet)                128            134          18          0.8        1282.2       0.3X
+100 deep x 1000 rows (read in-mem)                  345            350           5          0.3        3447.8       0.1X
+100 deep x 1000 rows (exec in-mem)                 1283           1283           0          0.1       12830.5       0.0X
+100 deep x 1000 rows (read parquet)                1201           1205           7          0.1       12005.2       0.0X
+100 deep x 1000 rows (write parquet)                436            443           9          0.2        4361.4       0.1X
+250 deep x 400 rows (read in-mem)                  1882           1883           1          0.1       18819.9       0.0X
```
Comment (MaxGekk) [inline sql/core/benchmarks/WideSchemaBenchmark-results.txt]: Regression ~8 times, https://issues.apache.org/jira/browse/SPARK-30564

File: sql/hive/benchmarks/OrcReadBenchmark-results.txt [hunk 9]
```diff
@@ -2,155 +2,155 @@
...
+OpenJDK 64-Bit Server VM 1.8.0_232-8u232-b09-0ubuntu1~18.04.1-b09 on Linux 4.15.0-1044-aws
 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
 Single Column Scan from 100 columns:      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-Native ORC MR                                       290            350         102          3.6         276.1       1.0X
-Native ORC Vectorized                               155            166          15          6.7         148.2       1.9X
-Hive built-in ORC                                   520            531           8          2.0         495.8       0.6X
+Native ORC MR                                       250            264          15          4.2         238.1       1.0X
+Native ORC Vectorized                               121            138          24          8.7         115.5       2.1X
+Hive built-in ORC                                  1761           1792          43          0.6        1679.3       0.1X
```
Comment (MaxGekk) [inline sql/hive/benchmarks/OrcReadBenchmark-results.txt]: Here there is a regression ~3 times. https://issues.apache.org/jira/browse/SPARK-30565

File: sql/core/benchmarks/IntervalBenchmark-results.txt (hunk 5 above)
Comment (srowen) [reply sql/core/benchmarks/IntervalBenchmark-results.txt]: Hm, regression vs what here... the change should have made the benchmarks faster, but they were also re-run. Is it the new instance type?

Comment (MaxGekk) [reply sql/core/benchmarks/IntervalBenchmark-results.txt]: There are a few commits in interval implementation between 2 re-runs. The regression can be related to those commits but to my changes.  

> Is it the new instance type?

It is the same.

**Comment by dongjoon-hyun** [comment]
Thank you for the follow-up, @MaxGekk .
